src/librustc/traits/error_reporting.rs

```diff
@@ -103,16 +105,191 @@ pub fn report_projection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
     }
 }
 
+fn impl_substs<'a, 'tcx>(fcx: &InferCtxt<'a, 'tcx>,
+                         did: DefId,
+                         obligation: PredicateObligation<'tcx>)
+                         -> subst::Substs<'tcx> {
+    let tcx = fcx.tcx;
+
+    let ity = tcx.lookup_item_type(did);
+    let (tps, rps, _) =
+        (ity.generics.types.get_slice(subst::TypeSpace),
+         ity.generics.regions.get_slice(subst::TypeSpace),
+         ity.ty);
+
+    let rps = fcx.region_vars_for_defs(obligation.cause.span, rps);
+    let mut substs = subst::Substs::new(
+        subst::VecPerParamSpace::empty(),
+        subst::VecPerParamSpace::new(rps, Vec::new(), Vec::new()));
+    fcx.type_vars_for_defs(obligation.cause.span, subst::ParamSpace::TypeSpace, &mut substs, tps);
+    substs
+}
+
+trait AssociatedWeight {
+    fn get_weight(&self) -> usize;
+}
+
+impl<'a> AssociatedWeight for TypeVariants<'a> {
+    // First number is for "global" weight and second number is for bigger precision
+    fn get_weight(&self) -> usize {
+        match *self {
+            TypeVariants::TyBool => 11,
+            TypeVariants::TyChar => 12,
+            TypeVariants::TyStr => 13,
+            TypeVariants::TyInt(_) => 21,
+            TypeVariants::TyUint(_) => 22,
+            TypeVariants::TyFloat(_) => 23,
+            TypeVariants::TyRawPtr(_) => 24,
+            TypeVariants::TyEnum(_, _) => 31,
+            TypeVariants::TyStruct(_, _) => 32,
+            TypeVariants::TyBox(_) => 33,
+            TypeVariants::TyTuple(_) => 34,
+            TypeVariants::TyArray(_, _) => 41,
+            TypeVariants::TySlice(_) => 42,
+            TypeVariants::TyRef(_, _) => 51,
+            TypeVariants::TyFnDef(_, _, _) => 52,
+            TypeVariants::TyFnPtr(_) => 53,
+            TypeVariants::TyTrait(_) => 61,
+            TypeVariants::TyClosure(_, _) => 71,
+            TypeVariants::TyProjection(_) => 81,
+            TypeVariants::TyParam(_) => 82,
+            TypeVariants::TyInfer(_) => 83,
+            TypeVariants::TyError => 91,
+        }
+    }
+}
+
+// The "closer" the types are, the lesser the weight
+fn get_weight_diff(a: &ty::TypeVariants, b: &TypeVariants, big_weight: bool) -> usize {
```

Comment:
nit: I'd introduce a new enum for `big_weight` rather than use a boolean here, so that I don't have to remember what `true` means at the call-sites. (E.g. `enum Weight { Big, Little }`.)